nexus/server.go

```diff
@@ -0,0 +1,497 @@
+	// Return an [UnsuccessfulOperationError] to indicate that an operation completed as failed or canceled.
+	StartOperation(context.Context, *StartOperationRequest) (OperationResponse, error)
+	// GetOperationResult handles requests to get the result of an asynchronous operation. Return
+	// [OperationResponseSync] to respond successfully - inline, or error with [ErrOperationStillRunning] to indicate
+	// that an asynchronous operation is still running.
+	// Return an [UnsuccessfulOperationError] to indicate that an operation completed as failed or canceled.
+	//
+	// When [GetOperationResultRequest.Wait] is greater than zero, this request should be treated as a long poll.
+	// Long poll requests have a server side timeout, configurable via [HandlerOptions.GetResultTimeout], and exposed
+	// via context deadline. The context deadline is decoupled from the application level Wait duration.
+	//
+	// It is the implementor's responsiblity to respect the client's wait duration and return in a timely fashion.
+	// Consider using a derived context that enforces the wait timeout when implementing this method and return
+	// [ErrOperationStillRunning] when that context expires as shown in the example.
+	GetOperationResult(context.Context, *GetOperationResultRequest) (*OperationResponseSync, error)
+	// GetOperationInfo handles requests to get information about an asynchronous operation.
+	GetOperationInfo(context.Context, *GetOperationInfoRequest) (*OperationInfo, error)
+	// CancelOperation handles requests to cancel an asynchronous operation.
+	// Cancelation in Nexus is:
+	//  1. asynchronous - returning from this method only ensures that cancelation is delivered, it may later be ignored
+	//  by the underlying operation implemention.
+	//  2. idempotent - implementors should ignore duplicate cancelations for the same operation.
+	CancelOperation(context.Context, *CancelOperationRequest) error
+	mustEmbedUnimplementedHandler()
+}
+
+// HandlerError is a special error that can be returned from [Handler] methods for failing an HTTP request with a custom
+// status code and failure message.
+type HandlerError struct {
+	// Defaults to 500.
+	StatusCode int
+	// Failure to report back in the response. Optional.
+	Failure *Failure
+}
+
+// Error implements the error interface.
+func (e *HandlerError) Error() string {
+	if e.Failure != nil {
+		return fmt.Sprintf("handler error (%d): %s", e.StatusCode, e.Failure.Message)
+	}
+	return fmt.Sprintf("handler error (%d)", e.StatusCode)
+}
+
+func newBadRequestError(format string, args ...any) *HandlerError {
+	return &HandlerError{
+		StatusCode: http.StatusBadRequest,
+		Failure: &Failure{
+			Message: fmt.Sprintf(format, args...),
+		},
+	}
+}
+
+type baseHTTPHandler struct {
+	logger *slog.Logger
+}
+
+type httpHandler struct {
+	baseHTTPHandler
+	options HandlerOptions
+}
+
+// CompletionRequest is input for CompletionHandler.CompleteOperation.
+type CompletionRequest struct {
```

Comment:
I'm wondering if the completion stuff here and on the client should go in a `completion.go` file. Servers don't need to concern themselves w/ this, so it's a bit strange to be in this file

Reply:
Sure, I'll move this as well as the client code.